### PR TITLE
support db_paths

### PR DESCRIPTION
**Pull request by follitude**
PTAL @zhangjinpeng1987

### CODE REVIEW

File: src/rocksdb_options.rs [hunk 1]
```diff
@@ -627,6 +659,30 @@ impl DBOptions {
             crocksdb_ffi::crocksdb_options_set_allow_concurrent_memtable_write(self.inner, v);
         }
     }
+
+    pub fn set_db_paths<P: Into<DbPath>>(&self, val: Vec<P>) {
```
Comment (siddontang) [inline src/rocksdb_options.rs]: is it better to use `&[P]` here? /cc @BusyJay

File: tests/test_rocksdb_options.rs [hunk 2]
```diff
@@ -452,3 +452,12 @@ fn test_clone_options() {
     let cf_opts2 = cf_opts.clone();
     assert_eq!(cf_opts.get_compression(), cf_opts2.get_compression());
 }
+
+#[test]
+fn test_db_paths() {
+    let path = TempDir::new("_rust_rocksdb_db_paths").expect("");
+    let mut opts = DBOptions::new();
+    opts.create_if_missing(true);
+    opts.set_db_paths(vec![("/home/test_db_path", 1073741824 as u64)]);
```
Comment (siddontang) [inline tests/test_rocksdb_options.rs]: do we still need the path when we pass `db_paths`?

Comment (follitude) [inline tests/test_rocksdb_options.rs]: yes, we still need.

File: librocksdb_sys/crocksdb/c.cc [hunk 3]
```diff
@@ -1992,6 +1993,18 @@ void crocksdb_options_set_use_fsync(
   opt->rep.use_fsync = use_fsync;
 }
 
+void crocksdb_options_set_db_paths(crocksdb_options_t *opt,
+                                   const char *const *dbpath_list,
+                                   const size_t *path_lens,
+                                   const uint64_t *target_size, int num_paths) {
+  std::vector<DbPath> db_paths;
+  for (int i = 1; i < num_paths; ++i) {
```
Comment (BusyJay) [inline librocksdb_sys/crocksdb/c.cc]: Why 1?

File: src/rocksdb_options.rs (hunk 1 above)
Comment (BusyJay) [inline src/rocksdb_options.rs]: Yes.

File: src/rocksdb_options.rs [hunk 4]
```diff
@@ -627,6 +659,30 @@ impl DBOptions {
...
+        let num_paths = val.len();
+        let paths = val.into_iter().map(|p| p.into()).collect::<Vec<_>>();
+        let mut cpaths = Vec::with_capacity(num_paths);
+        let mut cpath_lens = Vec::with_capacity(num_paths);
+        let mut sizes = Vec::with_capacity(num_paths);
+        for dbpath in &paths {
+            cpaths.push(dbpath.path
+                .to_str()
+                .map(|s| s.as_ptr() as _)
+                .unwrap_or_else(ptr::null));
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: This may causes undefined behavior.

File: src/rocksdb_options.rs [hunk 5]
```diff
@@ -158,6 +160,36 @@ impl UnsafeSnap {
     }
 }
 
+pub struct DbPath {
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: I don't think this struct is necessary.

File: src/rocksdb_options.rs [hunk 6]
```diff
@@ -627,6 +628,30 @@ impl DBOptions {
             crocksdb_ffi::crocksdb_options_set_allow_concurrent_memtable_write(self.inner, v);
         }
     }
+
+    pub fn set_db_paths(&self, val: &[(PathBuf, u64)]) {
```
Comment (siddontang) [inline src/rocksdb_options.rs]: what does u64 here mean? You should define a structure or a type for the tuple.

File: src/rocksdb_options.rs [hunk 7]
```diff
@@ -627,6 +628,31 @@ impl DBOptions {
             crocksdb_ffi::crocksdb_options_set_allow_concurrent_memtable_write(self.inner, v);
         }
     }
+
+    /// the second parameter is a slice which contains tuples (path, target_size).
+    pub fn set_db_paths(&self, val: &[(PathBuf, u64)]) {
+        let num_paths = val.len();
+        let paths: Vec<(PathBuf, u64)> = val.to_vec();
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: I don't think `to_vec` is necessary.

File: src/rocksdb_options.rs [hunk 8]
```diff
@@ -627,6 +628,31 @@ impl DBOptions {
             crocksdb_ffi::crocksdb_options_set_allow_concurrent_memtable_write(self.inner, v);
         }
     }
+
+    /// the second parameter is a slice which contains tuples (path, target_size).
+    pub fn set_db_paths(&self, val: &[(PathBuf, u64)]) {
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Use `Path` instead.

File: src/rocksdb_options.rs [hunk 9]
```diff
@@ -627,6 +628,31 @@ impl DBOptions {
...
+        let num_paths = val.len();
+        let paths: Vec<(PathBuf, u64)> = val.to_vec();
+        let mut cpaths = Vec::with_capacity(num_paths);
+        let mut cpath_lens = Vec::with_capacity(num_paths);
+        let mut sizes = Vec::with_capacity(num_paths);
+        for dbpath in &paths {
+            cpaths.push(dbpath.0
+                .to_str()
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: `unwrap().as_ptr()` looks better.

File: src/rocksdb_options.rs [hunk 10]
```diff
@@ -627,6 +628,31 @@ impl DBOptions {
...
+                .map(|s| s.as_ptr() as _)
+                .unwrap());
+            cpath_lens.push(dbpath.0.to_str().map(|s| s.len()).unwrap());
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Save `to_str` as a temporary variable.

File: tests/test_rocksdb_options.rs [hunk 11]
```diff
@@ -452,3 +452,12 @@ fn test_clone_options() {
     let cf_opts2 = cf_opts.clone();
     assert_eq!(cf_opts.get_compression(), cf_opts2.get_compression());
 }
+
+#[test]
+fn test_db_paths() {
+    let path = TempDir::new("_rust_rocksdb_db_paths").expect("");
+    let mut opts = DBOptions::new();
+    opts.create_if_missing(true);
+    opts.set_db_paths(&[(PathBuf::from("rust_rocksdb_test_db_path"), 1073741824 as u64)]);
```
Comment (BusyJay) [inline tests/test_rocksdb_options.rs]: Will this directory be created and cleaned up when tests are finished?

File: src/rocksdb_options.rs [hunk 12]
```diff
@@ -614,6 +615,28 @@ impl DBOptions {
...
+    /// the second parameter is a slice which contains tuples (path, target_size).
+    pub fn set_db_paths<T: AsRef<Path>>(&self, val: &[(T, u64)]) {
+        let num_paths = val.len();
+        let mut cpaths = Vec::with_capacity(num_paths);
+        let mut cpath_lens = Vec::with_capacity(num_paths);
+        let mut sizes = Vec::with_capacity(num_paths);
+        for dbpath in val {
+            let dbpath_str = dbpath.0.as_ref().to_str();
+            cpaths.push(dbpath_str.unwrap().as_ptr() as _);
+            cpath_lens.push(dbpath_str.map(|s| s.len()).unwrap());
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Use `unwrap().len()` instead

Comment (follitude) [inline src/rocksdb_options.rs]: ok

**Comment by andelf** [comment]
LGTM

**Pull request merged**